core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolverGroup.java

```diff
@@ -0,0 +1,173 @@
+ *  version 2.0 (the "License"); you may not use this file except in compliance
+ *  with the License. You may obtain a copy of the License at:
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  License for the specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.NetworkInterface;
+import java.net.SocketException;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Consumer;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+
+import com.linecorp.armeria.client.RefreshingAddressResolver.CacheEntry;
+import com.linecorp.armeria.client.retry.Backoff;
+import com.linecorp.armeria.internal.dns.DefaultDnsNameResolver;
+
+import io.netty.channel.EventLoop;
+import io.netty.handler.codec.dns.DnsRecordType;
+import io.netty.resolver.AddressResolver;
+import io.netty.resolver.AddressResolverGroup;
+import io.netty.resolver.ResolvedAddressTypes;
+import io.netty.resolver.dns.DnsNameResolverBuilder;
+import io.netty.util.NetUtil;
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * Creates and manages refreshing {@link AddressResolver}s and the DNS cache.
+ */
+final class RefreshingAddressResolverGroup extends AddressResolverGroup<InetSocketAddress> {
+
+    private static final Logger logger = LoggerFactory.getLogger(RefreshingAddressResolverGroup.class);
+
+    private static final List<DnsRecordType> defaultDnsRecordTypes;
+
+    static {
+        final ResolvedAddressTypes resolvedAddressTypes;
+        if (NetUtil.isIpV4StackPreferred() || !anyInterfaceSupportsIpV6()) {
+            resolvedAddressTypes = ResolvedAddressTypes.IPV4_ONLY;
+        } else {
+            if (NetUtil.isIpV6AddressesPreferred()) {
+                resolvedAddressTypes = ResolvedAddressTypes.IPV6_PREFERRED;
+            } else {
+                resolvedAddressTypes = ResolvedAddressTypes.IPV4_PREFERRED;
+            }
+        }
+
+        defaultDnsRecordTypes = dnsRecordTypes(resolvedAddressTypes);
+    }
+
+    /**
+     * Returns {@code true} if any {@link NetworkInterface} supports {@code IPv6}, {@code false} otherwise.
+     */
+    private static boolean anyInterfaceSupportsIpV6() {
```

Comment:
I remember when this was added to Netty ;) Could you mark any code copied from Netty? We should work on making them public or otherwise avoiding keeping copied code forever and missing out on upstream bug fixes. I think a lot of our DNS code is copied so we can probably help improve the abstraction in Netty itself.

Reply:
Right. 😃 Let me see what I can do. Thanks!